### ISSUE TITLE
Training stops when logging images to wandb in DDP mode

### ISSUE THREAD

**Issue by markovalexander** [huggingface/accelerate#785]
### System Info

```Shell
- `Accelerate` version: 0.13.2
- Platform: Linux-5.4.0-42-generic-x86_64-with-glibc2.10
- Python version: 3.8.12
- Numpy version: 1.21.4
- PyTorch version (GPU?): 1.11.0a0+b6df043 (True)
- `Accelerate` default config:
        - compute_environment: LOCAL_MACHINE
        - distributed_type: MULTI_GPU
        - mixed_precision: no
        - use_cpu: False
        - num_processes: 2
        - machine_rank: 0
        - num_machines: 1
        - gpu_ids: 2,3
        - main_process_ip: None
        - main_process_port: None
        - rdzv_backend: static
        - same_network: True
        - main_training_function: main
        - deepspeed_config: {}
        - fsdp_config: {}
        - downcast_bf16: no
```

### Information

- [ ] The official example scripts
- [X] My own modified scripts

### Tasks

- [ ] One of the scripts in the examples/ folder of Accelerate or an officially supported `no_trainer` script in the `examples` folder of the `transformers` repo (such as `run_no_trainer_glue.py`)
- [X] My own task or dataset (give details below)

### Reproduction

I am trying to train ViT-VQGAN using accelerate, and log reconstructed images during training to wandb. I used 

    if params.training_params.report_to_wandb:
        accelerator.init_trackers('vit_vqgan', {'config_name': config, 'lr': lr, 'params': params.dict()})

before training loop. Next, goes training loop itself:

    progress_bar = tqdm(total=len(train_loader) * num_epochs, disable=not accelerator.is_local_main_process)

    for _ in range(num_epochs):
        for step_idx, batch in enumerate(train_loader):
            model.train()
            with accelerator.accumulate(model), accelerator.autocast():
                images = batch[0]
                global_step += 1
                step_type = StepType.from_global_step(global_step)

                model_output: ViTVQGANOutput = model(images, step_type, global_step, step_idx)

                loss = model_output.loss

                accelerator.backward(loss)

                if step_type.AUTOENCODER:
                    optimizer_model.step()
                    optimizer_model.zero_grad()
                else:
                    optimizer_loss.step()
                    optimizer_loss.zero_grad()

                if params.training_params.report_to_wandb and \
                    global_step % params.training_params.log_steps == 0:
                    accelerator.log(
                        {'train_loss': loss, 'quantizer_loss': model_output.quantizer_loss}, step=global_step,
                    )
            progress_bar.update(1)

            if global_step % params.training_params.eval_steps == 0 and params.training_params.report_to_wandb:
                model.eval()
                with accelerator.main_process_first():
                    with torch.inference_mode():
                        reconstructed = model(test_images).reconstructed

                    input_image = make_grid(test_images, nrow=2)
                    reconstructed = make_grid(reconstructed, nrow=2)

                    input_log = wandb.Image(
                        input_image.detach().cpu().numpy().transpose((1, 2, 0)), caption='input image',
                    )
                    reconstructed_log = wandb.Image(
                        reconstructed.detach().cpu().numpy().transpose((1, 2, 0)), caption='reconstructed',
                    )

                    wandb.log({'input image': input_log, 'reconstructed image': reconstructed_log})

In my case, I log loss every 100 steps and images every 500 steps. However, after images were pushed to wandb:
<img width="986" alt="image" src="https://user-images.githubusercontent.com/22663468/197392750-696d3fb8-5a7c-496c-8a8c-895b7c54234f.png">

the training freezes (i assume deadlock, but i am not sure) and i see no progress on progress bar and no new logs in wandb.

### Expected behavior

```Shell
Training continues
```

**Comment by markovalexander** [comment]
I added `accelerator.print('left main_process_first()')` after leaving `accelerator.main_process_first()` сontext manager and i can see this print statement: 
<img width="1208" alt="image" src="https://user-images.githubusercontent.com/22663468/197393501-7560b6fa-b47d-4ca8-b29c-5bf1b6a9b415.png">

and it stopped...

**Comment by sgugger** [comment]
cc @muellerzr

**Comment by pacman100** [comment]
Hello @markovalexander, wand is only available for main process. Can you try gathering the reconstructed and original images on main process via 
```
reconsrtucted_images_gathered, original_images_gathered =  accelerator.gather((reconstructed, test_images))
```
Post that make grids and log to wandb using 
```
input_image = make_grid(original_images_gathered, nrow=2)
reconstructed = make_grid(reconsrtucted_images_gathered, nrow=2)

input_log = wandb.Image(
    input_image.detach().cpu().numpy().transpose((1, 2, 0)), caption='input image',
)
reconstructed_log = wandb.Image(
    reconstructed.detach().cpu().numpy().transpose((1, 2, 0)), caption='reconstructed',
)
accelerator.get_tracker("wandb").log({'input image': input_log, 'reconstructed image': reconstructed_log})
``` 

Could you please try above changes and let us know if it addresses your issue?

**Comment by markovalexander** [comment]
@pacman100 hello! Thank you for your advice, it actually solved my problem! (I also found out that I forgot to move `test_images`, but I checked that it is not enough and `accelerator.gather` is required)
